iOS cli option

Debusan13 opened the pull request:
# Adding full support for iOS backups
## The Problem
So I noticed in multiple reddit threads people were asking you if this tool could be used on iOS backups.[[1]](https://old.reddit.com/r/DataHoarder/comments/108gnnh/i_wanted_to_be_able_to_exportbackup_imessage/j3tu4ss/) [[2]](https://old.reddit.com/r/DataHoarder/comments/108gnnh/i_wanted_to_be_able_to_exportbackup_imessage/)
I also wanted to do this, but the first issue is "where is the `chat.db` file?" How does one extract the chat info, attachments, etc when the backup has nonsense hashed file names? Building off of this, how can we retain most of the core code while still allowing backups from iOS? It shouldn't be tedious either and should work from any backup location as people might use Finder, iMazing, or an external drive which all store the backup in different locations.

## Finding `chat.db`
From my limited testing and thanks to @nprezant's work 5 years ago, it seems that iOS backups keep their chat.db file in the same location: `$PATH_TO_BACKUP/3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28`. After finding this I did a test run and found that the messages backed up no problem **but there were no attachments**
<p align="center">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/47309906/234127303-26f3b2b6-0244-44e6-a9d0-31fe1b067d9d.png" >
</p>

Notice that all the file paths are paths from the iOS system, meaning that all the attachments are stored somewhere in the backup, they are just hidden by the hash.

##### If all you want is texts from your iPhone run the following
```sh
imessage-exporter -n -f html -p $PATH_TO_BACKUP/3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28 -o $EXPORT_DIR
```

### Finding 'contacts.db'
One thing I also found that might prove helpful for issues like #61 is a `contacts.db` stored at: `$PATH_TO_BACKUP/31/31bb7ba8914766d4ba40d6dfb6113c8b614be442`

Since the hash is a consistent pattern, these two file locations _shouldn't change_, but there is no guarantee.

## The Hash
How the files are hashed from "original path" -> "hashed path" is as follows:
```python
def get_hashed_path(raw_path):
    domainpath = 'MediaDomain-' + raw_path[2:] # create the domain path that will be hashed
    filename = sha1(domainpath.encode('utf-8')).hexdigest() # create the SHA1 hash (which is the file name)
    dir = filename[:2] # the first two digets of the filename are the directory name
    return Path(dir, filename) # create an file path to the attachment
```
so for example,
```console
$ python3 get_hashed_path('~/Library/SMS/Attachments/40/00/9D1D2F2B-FFF9-4791-85A3-AC609E584FF4/IMG_4567.jpeg')
PosixPath('c7/c7c61670bfead910aa7c066be22b14e17ef793c7')
```
<sup>Thanks again to [nprazent](https://github.com/nprezant/iMessageBackup/blob/940d001fb7be557d5d57504eb26b3489e88de26e/imessage_backup_tools.py#L82)</sup>

The `SHA1` hash requires the crate [sha1 v0.10.5](https://crates.io/crates/sha1) to be added to the `Cargo.toml' 
3aa16204acdbe8d7659cbef4a4680f7b2f56572c

https://github.com/ReagentX/imessage-exporter/blob/3aa16204acdbe8d7659cbef4a4680f7b2f56572c/imessage-exporter/Cargo.toml#L19

## `--ios` CLI Argument
Since the backup exists in its own little world, and I didn't want to rewrite a lot of the reader and exporter, I thought the best option was to add a `--ios` flag which requires the `-p, --db-path` to point to the folder holding the iOS backup. So this would be `Library/Application\ Support/MobileSync/Backup/<backup-version>`

This is what the help dialogue says for the option
```
        --ios
            Specify that the database is from an iOS backup
            Using this option requires a custom path to the iPhone backup directory
```
https://github.com/ReagentX/imessage-exporter/blob/592973c31c7f443a6e7ece917ad0d6f8e2e0addf/imessage-exporter/src/app/options.rs#L39-L58
https://github.com/ReagentX/imessage-exporter/blob/592973c31c7f443a6e7ece917ad0d6f8e2e0addf/imessage-exporter/src/app/options.rs#L71
https://github.com/ReagentX/imessage-exporter/blob/8b25dc288dc68e5a3da44233324c041f6f0abbfe/imessage-exporter/src/app/options.rs#L286-L292

### Validating the path
I found that the best way to validate the path was to check for the aforementioned [chat.db](#####If-all-you-want-is-texts-from-your-iPhone-run-the-following).
https://github.com/ReagentX/imessage-exporter/blob/592973c31c7f443a6e7ece917ad0d6f8e2e0addf/imessage-exporter/src/app/options.rs#L119-L127
The value for the path is a public constant, and while I was at it I added the contact.db file location as well in case it's used later, even though it is unused code
https://github.com/ReagentX/imessage-exporter/blob/592973c31c7f443a6e7ece917ad0d6f8e2e0addf/imessage-exporter/src/app/options.rs#L14-L17

### Establishing the database connection
Since the iOS Backup lives in its own directory, the path of the `chat.db` needs to be appended to the path to the iOS backup. However, if the iOS option is not enabled, then the path should remain the same.
https://github.com/ReagentX/imessage-exporter/blob/ae0b29b1f29eae5c51f627d85af698a530950870/imessage-exporter/src/app/runtime.rs#L168-L174

## HTML Exporter
Since the HTML Exporter already has `resolved_attachment_path` to resolve the relative path, simply adding a control flow statement to determine which path is used was all that was necessary. If the `ios` flag is true, then the path to the attachment from the database will be rehashed to find its location in the backup. This seemed to work perfectly and attachments were appearing in the html files no problems.
https://github.com/ReagentX/imessage-exporter/blob/034c90878a22b6c6fd8ba62110556b8f9d9ed120/imessage-exporter/src/exporters/html.rs#L404-L415
> **Note**
> Using `.display()` might yield some issues since it ["may perform lossy conversion"](https://doc.rust-lang.org/std/path/struct.Path.html#method.display) meaning it won't display non-UTF-8 characters. Not sure how much of a problem this really is.

The exporter properly creates the `Attachments` folder, where each subdirectory uses the `chat_id` and each file name is the "original" file name and not the "hashed" one.
<img width="580" alt="image" src="https://user-images.githubusercontent.com/47309906/234141735-8932d544-e40d-426d-b340-e207e9b6ddb8.png">

## TXT Exporter
I thought it fitting I should still update the TXT Exporter using these file paths, as the ones it would display were files that did not exist on the machine.
<p align=center>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/47309906/234140242-91d92a8d-6c03-4437-a751-b63be3a88304.png">
</p>
I made it so the file path actually points to the file's location in the iOS backup, and whats more I formatted it such that it could be easily pasted into the terminal using `cat` or `bat` allowing someone to have the file in its correct format.
<p align=center>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/47309906/234142761-0d3a4765-aeb8-437d-b259-a3ea2c978791.png">
<img width="400" alt="image" src="https://user-images.githubusercontent.com/47309906/234143044-cc891703-edf9-44c8-aa4b-92a17a2d3715.png">
</p?>

## Possible improvements

- Path concatenation and creation done in a more "rusty" way, I am not that experienced and I'm not sure if what I did was proper
- `--ios` arugment location, I put it at the end, but probably having it appear closer to the front would be better
   - making the help section more clear that the `-p` argument should be used with it
 - removing escape slashes on paths when adding directories. macOS has the feature to drag and drop directories into the terminal however doing so includes the escape slash in front of any whitespace `\ `. I made sure to **add**  that back in for the TXT export, but having the option to shave it off would be nice as I am sure many people will "drop" their iOS backup directory into the terminal and not know they have to remove the escape character for the path to read properly

Thanks again to @nprezant and their repo [iMessageBackup](https://github.com/nprezant/iMessageBackup)
and yes the emojis are non negotiable 😜

Debusan13 committed:
Removed nested match statements 🪺

Debusan13 committed:
added "ios" option to args $_📱
set to false for test cases

Debusan13 committed:
added sha1 hasher dependency #️⃣🧂🔐

Debusan13 committed:
ios requires OPTION_DB_PATH ⚠️⛓️

Debusan13 committed:
modified db path for conn based on ios 📀🥞🔌

Debusan13 committed:
ios option no longer takes argument 📲🚫⛳️

Debusan13 committed:
html reads hashed filepath!! 😎💪📖#️⃣

Debusan13 committed:
enhanced filepath for txt output ✨📄📟

Debusan13 committed:
Escape spaces in path for easier copy-paste🔍🛣️🚪

ReagentX commented on imessage-exporter/src/exporters/txt.rs:
```diff
@@ -276,9 +278,27 @@ impl<'a> Writer<'a> for TXT<'a> {
+                        .unwrap_or_else(|| self.config.options.db_path.display().to_string());
+                    // <db_path>/<hash[0..2]>/<hash>/<filename> > <filename> allows for copy-paste conversion
+                    format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash,
+                                             filename.rsplit_once("/").unwrap().1)
```
Please don't use `unwrap()` in runtime code; I try to reserve it for tests only.

Debusan13 replied:
Gotcha, would this be better?
```rust
format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash,
        filename.rsplit_once("/").map(|(_, x)| x).unwrap_or_else(|| "Could not get filename")
```
or we could reflow the logic so it goes ios, macos, error in the match
```rust
match &attachment.filename {
  Some(filename) if self.config.options.ios => {
      // iOS rehash the filename to a path in the backup
      let salt = "MediaDomain-";
      let hash = format!("{:x}", Sha1::digest(
          format!("{}{}", salt, &filename[2..]).as_bytes()
      ));
      // attempting to escape spaces in the path for easier copy-paste
      let db_path_os_string = self.config.options.db_path.to_str()
          .map(|s| s.replace(" ", r#"\ "#))
          .unwrap_or_else(|| self.config.options.db_path.display().to_string());
      // <db_path>/<hash[0..2]>/<hash>/<filename> > <filename> allows for copy-paste conversion
      Ok(format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash,
                 filename.rsplit_once("/").map(|(_, x)| x).unwrap_or_else(|| "Could not get filename")))
  }
  // macOS uses the filename as the path
  Some(filename) => Ok(filename.to_string()), 
  None => Err(attachment.filename()),
}
```
and we could always swap out the "Could not get a filename" with just a plain old error like so
```rust
match filename.rsplit_once("/") {
    Some((_, filename)) => Ok(format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash, filename)),
    None => Err(attachment.filename()),
}
```
or returning it "macOS" style, which is the path from the internal iPhone file system
```rust
match filename.rsplit_once("/") {
    Some((_, filename)) => Ok(format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash, filename)),
    None => Ok(filename.to_string()), 
}
```

again I'm not entirely sure how you want the program to handle it or what syntactic style you prefer

ReagentX replied:
1) I think the logic for the iOS parsing should be split out somewhere so that it can be tested separately. 

2) I think this whole block will be a lot cleaner once we use an enum instead of a `bool` here, so I will wait to review it until that changes.

ReagentX commented:
Really cool work, thank you for the contribution. This is the most detailed pull request I have ever seen!

I will take a closer look soon; I left a small comment in the meantime.

ReagentX commented on imessage-exporter/src/app/runtime.rs:
````diff
@@ -166,7 +166,12 @@ impl<'a> Config<'a> {
     /// ```
     pub fn new(options: Options) -> Result<Config, RuntimeError> {
-        let conn = get_connection(&options.db_path).map_err(RuntimeError::DatabaseError)?;
+        let conn = if options.ios {
````
Instead of storing `iOS` as a Boolean, we should create an enum like `ExportSource` that has variants for `MacOS` and `iOS` so we may pattern match on them later should we expand this logic in the future.

Debusan13 replied:
Yes I agree, I wasn't sure how to go about it since I was largely going off of what I saw, is there a design doc or class diagram I could refer to?

ReagentX replied:
https://doc.rust-lang.org/rust-by-example/custom_types/enum.html

Debusan13 replied:
Yes I understand the basic concept, what I think of though is a relatively simple implementation 
```rust
enum ExportPlatform {
    MacOS,
    IOS,
}

match export_platform {
    ExportPlatform::MacOS => {
        // macOS-specific export code
    }
    ExportPlatform::IOS => {
        // iOS-specific export code
    }
}
```

I meant where in the program would I put this enum? Like is there a specific hierarchy or structure you use with your "classes" (outdated term) that I could follow, or am I free to put in anywhere I see fit?

ReagentX replied:
That implementation looks reasonable to me. It should live somewhere in the library if it is used in the library, otherwise it should live in the binary's `options.rs` for now.

ReagentX commented on imessage-exporter/src/exporters/html.rs:
```diff
@@ -399,12 +401,18 @@ impl<'a> Writer<'a> for HTML<'a> {
+                        // Resolve the attachment path for iOS backup
+                        } else if self.config.options.ios {
+                            let salt = "MediaDomain-";
+                            let hash = format!("{:x}", Sha1::digest(format!("{}{}", salt, &path_str[2..]).as_bytes()));
```
Use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get, don't directly index as it can panic.

ReagentX commented on imessage-exporter/src/exporters/html.rs:
```diff
@@ -399,12 +401,18 @@ impl<'a> Writer<'a> for HTML<'a> {
+                        } else if self.config.options.ios {
+                            let salt = "MediaDomain-";
+                            let hash = format!("{:x}", Sha1::digest(format!("{}{}", salt, &path_str[2..]).as_bytes()));
+                            format!("{}/{}/{}", self.config.options.db_path.display(), &hash[0..2], &hash)
```
Same as above

ReagentX commented on imessage-exporter/src/exporters/txt.rs:
```diff
@@ -276,9 +278,27 @@ impl<'a> Writer<'a> for TXT<'a> {
+                        .map(|s| s.replace(" ", r#"\ "#))
+                        .unwrap_or_else(|| self.config.options.db_path.display().to_string());
+                    // <db_path>/<hash[0..2]>/<hash>/<filename> > <filename> allows for copy-paste conversion
+                    format!("{}/{}/{} > {}", db_path_os_string, &hash[0..2], hash,
```
Use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get, don't directly index as it can panic.

ReagentX commented on imessage-exporter/src/exporters/txt.rs:
```diff
@@ -276,9 +278,27 @@ impl<'a> Writer<'a> for TXT<'a> {
+                    // iOS rehash the filename to a path in the backup
+                    let salt = "MediaDomain-";
+                    let hash = format!("{:x}", Sha1::digest(
+                        format!("{}{}", salt, &filename[2..]).as_bytes()
```
Use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get, don't directly index as it can panic.

ReagentX commented on imessage-exporter/src/app/options.rs:
```diff
@@ -11,6 +11,10 @@ use crate::app::error::RuntimeError;
+/// Default location in an iOS backup to find the iMessage database
+pub const DEFAULT_IOS_CHATDB_PATH: &str = "3d/3d0d7e5fb2ce288813306e4d4636395e047a3d28";
+/// Default location in an iOS backup to find the iMessage contacts database
+pub const DEFAULT_IOS_CONTACTSDB_PATH: &str = "31/31bb7ba8914766d4ba40d6dfb6113c8b614be442"; // unused
```
Remove this

ReagentX commented on imessage-exporter/src/app/options.rs:
```diff
@@ -108,6 +116,18 @@ impl<'a> Options<'a> {
+            let db_path = PathBuf::from(user_path.unwrap());
+            if !db_path.join(DEFAULT_IOS_CHATDB_PATH).exists() {
+                return Err(RuntimeError::InvalidOptions(format!(
+                    "Option {OPTION_IOS} is enabled, but the database path does not appear to be a valid iOS backup"
```
This message should accurately reflect what went wrong here: it's not invalid, the file was not found at the path specified.

Debusan13 committed:
Removed contactdb path 🚫🗑️

Debusan13 committed:
ios cli error message more accurate ⚠️💬✅

ReagentX commented:
If it makes things easier, I can merge this current branch into another feature branch and make all of these fixes myself. 

You've done a lot of work here; since I'm going to reorganize some of this anyway I may as well also build these requested fixes. If you just wrap up the enum aspect, I will merge this and clean the rest up in my own way.

Debusan13 commented:
I fixed up a lot of the little things (the unwraps, removing the unused constant, made the ios error message more clear) and im currently in progress on the enum. Ill push everything that I have by the end of today (~4:00 AM Wednesday (GMT)).

Debusan13 committed:
ImportPlatform enum created, added to Options 🏗️💻📱

Debusan13 committed:
from_args updated to use ImportPlatform enum 😇👞

Debusan13 committed:
Options::get_db_path created, initializes conn🧶🪛
also updated runtime tests to include ImportPlatform enum,
but they remain unused

Debusan13 committed:
ImportPlatform::get_attachment_path_txt created 🔨
using in txt.rs
left very verbose in for changes

Debusan13 committed:
moved "resolve file path" to ImportPlatform
not entirely sure how errors should be handled

Debusan13 commented:
Alright, I took a swing at creating the enum and impl for it. Hopefully it makes the reorganizing easier for you. I would love to see how this feature ends up looking in the end as this was my first time writing "production" rust. Thank you for your help and insight, I learned a lot!

Debusan13 committed:
one last get_db_path() update 😎💪🥅

Debusan13 committed:
fixed function calls 📞😵‍💫

ReagentX commented:
This is really good work, you should be proud of it!
